### PR TITLE
[Dev][Python] Add nightly test to execute all sqllogic tests using the python package

### PR DESCRIPTION
**Pull request by Tishj**
### Python SQLLogic Library

The library consists of a Parser that creates `SQLLogicTest` objects containing the parsed statements, and an executor that takes these parsed statements and uses the duckdb python package to execute them.

Using this executor as an example/base the goal is to use this to write executors/code generators for other languages in the future.
For example `pyodbc` could be used to implement an executor for our ODBC client.

### NightlyTest

A test is added to the nightly test suite that builds duckdb python on python 3.9 and runs the executor.
Note this is built without extensions, so the tests requiring extensions are skipped.

### CODE REVIEW

File: test/sqlite/sqllogic_command.cpp [hunk 1]
```diff
@@ -195,6 +195,8 @@ void LoopCommand::ExecuteInternal(ExecuteContext &context) const {
 	loop_def.loop_idx = definition.loop_start;
 	if (loop_def.is_parallel) {
 		for (auto &running_loop : context.running_loops) {
+			// FIXME: this is only checking one level down
+			// concurrent loop followed by a regular loop, with another concurrent loop inside is not prevented by this
```
Comment (Mytherin) [inline test/sqlite/sqllogic_command.cpp:199]: Could you show a test case where that happens? This is looping over all running loops, and trying this out seems to be working as intended:

###### test/test.test
```sql
concurrentloop i 0 100

loop j 0 100

concurrentloop k 0 100


endloop

endloop



endloop


```bash
build/debug/test/unittest test/test.test
> explicitly with message:
  Failure at :-1094795586: Nested parallel loop commands not allowed

```

(although there seems to be an issue with an uninitialized value here)

**Review by Mause — CHANGES_REQUESTED** [review]

File: scripts/sqllogictest/result.py [hunk 2]
```diff
@@ -0,0 +1,1256 @@
...
+        self.add_keyword(key, res)
+
+    def get_loop_statements(self):
+        saved_iterator = self.iterator
+        # Loop until EndLoop is found
+        statement = None
+        depth = 0
+        while self.iterator < len(self.statements):
+            statement = self.next_statement()
+            if statement.__class__ in [Foreach, Loop]:
```
[the remark this answers is not in the document]
Comment (Tishj) [reply scripts/sqllogictest/result.py:1079]: I'm not sure what the difference in behavior would be from the current solution?
These are types we create in the parser, nothing should derive from this so `isinstance` is not necessary afaik
And I'm not sure what the difference is between `type(...)` and `....__class__` ?

File: scripts/sqllogictest/result.py [hunk 3]
```diff
@@ -0,0 +1,1256 @@
...
+        'connection',
+        'cursors',
+    ]
+
+    def __init__(self, con: duckdb.DuckDBPyConnection):
+        assert con
+        self.cursors = {}
+        self.connection = con
+
+    def __del__(self):
```
[the remark this answers is not in the document]
Comment (Tishj) [reply scripts/sqllogictest/result.py:341]: Yea that's the idea, I'm not sure if this is entirely necessary in the end
But we share the same `SQLLogicConnectionPool` between (nested) Contexts.

In the end what is necessary is the `gc.collect()` call, to ensure that DatabaseInstance's are properly closed so they are removed from the `DBInstanceCache`.

Some tests share the same `load <db_name>` calls, so in those cases it's important that the in-memory version is cleaned up, we already make sure to delete the files after finishing the test, but if the in-memory version isn't removed that will be used instead.

File: scripts/sqllogictest/expected_result.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+from enum import Enum, auto
+from typing import Optional, List
+
+
+class ExpectedResult:
+    class Type(Enum):
+        SUCCESS = (auto(),)
```
[the remark this answers is not in the document]
Comment (Tishj) [reply scripts/sqllogictest/expected_result.py:7]: Euhm, good question, I think just a typo
Actually what I really wanted here was not explicitly assigning a value to the enum constants, only using them for their name

File: test/sqlite/sqllogic_command.cpp (hunk 1 above)
Comment (Tishj) [reply test/sqlite/sqllogic_command.cpp:199]: Ah I see now, the `running_loops` contains all the levels of nesting, so yes the comment is wrong, all levels are checked 👍

Comment (Tishj) [reply test/sqlite/sqllogic_command.cpp:199]: I think I misread and mistakenly thought this was checking the inner loops when we encountered a `loop` / `concurrent loop` command:

```sql
concurrent loop <loop we're checking>

loop <inner 1>

endloop

loop <inner 2>

endloop

loop endloop <inner 3>

endloop
```

But that's not what's happening, we're checking the parent loops instead

File: .github/workflows/NightlyTests.yml [hunk 5]
```diff
@@ -310,6 +310,47 @@ jobs:
        run: |
          ${GITHUB_WORKSPACE}/debug-build/bin/python3 -m pytest tools/pythonpkg/tests/fast --skiplist TestMemoryLeaks
 
+  python-sqllogic:
+     name: Python SQLLogicTest Library (Linux)
+     runs-on: ubuntu-latest
+     needs: linux-memory-leaks
+     container: ubuntu:18.04
```
Comment (Mytherin) [inline .github/workflows/NightlyTests.yml:317]: Can we move this outside of an old Ubuntu VM and instead just run on `ubuntu-latest`?

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**